server/queue/queue-pulsar/src/main/scala/org/apache/james/queue/pulsar/PulsarMailQueue.scala

```diff
@@ -0,0 +1,600 @@
+import scala.jdk.CollectionConverters._
+import scala.jdk.DurationConverters._
+import scala.math.Ordered.orderingToOrdered
+
+private[pulsar] object serializers {
+  implicit val headerFormat: Format[Header] = Json.format[Header]
+  implicit val enqueueIdFormat: Format[EnqueueId] = new Format[EnqueueId] {
+    override def writes(o: EnqueueId): JsValue = JsString(o.value)
+
+    override def reads(json: JsValue): JsResult[EnqueueId] =
+      json.validate[String].map(EnqueueId.apply).flatMap(_.fold(JsError.apply, JsSuccess(_)))
+  }
+  implicit val mailMetadataFormat: Format[MailMetadata] = Json.format[MailMetadata]
+}
+
+private[pulsar] object schemas {
+  implicit val schema: Schema[String] = Schema.STRING
+}
+
+/**
+ * In order to implement removal of mails from the queue, `PulsarMailQueue` makes use of a topic
+ * in which some filters are pushed. That way, all instances of `PulsarMailQueue` in a cluster
+ * eventually start dropping mails matching filters, effectively removing them from mail processing.
+ *
+ * The filtering is handled by a `FilterStage` Actor which maintains a set of active filters published by
+ * the `remove` method. It is responsible for dropping filters that can no longer match any message, and providing
+ * a consistent view of the messages that will be processed to the `browse` method.
+ *
+ * A filter cannot remove messages that are enqueued after the call to the `remove` method.
+ */
+class PulsarMailQueue(
+  name: MailQueueName,
+  config: PulsarConfiguration,
+  blobIdFactory: BlobId.Factory,
+  mimeMessageStore: Store[MimeMessage, MimeMessagePartsId],
+  mailQueueItemDecoratorFactory: MailQueueItemDecoratorFactory,
+  metricFactory: MetricFactory,
+  gaugeRegistry: GaugeRegistry,
+  system: ActorSystem
+) extends MailQueue with ManageableMailQueue {
+
+  import schemas._
+  import serializers._
+
+  type MessageAsJson = String
+
+  private val enqueueBufferSize = 10
+  private val requeueBufferSize = 10
+  private val awaitTimeout = 10.seconds
+
+  gaugeRegistry.register(QUEUE_SIZE_METRIC_NAME_PREFIX + name, () => getSize)
+  private val dequeueMetrics = metricFactory.generate(DEQUEUED_METRIC_NAME_PREFIX + name.asString)
+  private val enqueueMetric = metricFactory.generate(ENQUEUED_METRIC_NAME_PREFIX + name.asString)
+
+  private implicit val implicitSystem: ActorSystem = system
+  private implicit val ec: ExecutionContextExecutor = system.dispatcher
+  private implicit val implicitBlobIdFactory: BlobId.Factory = blobIdFactory
+  private implicit val client: PulsarAsyncClient = PulsarClient(config.brokerUri)
+  private val admin = {
+    val builder = PulsarAdmin.builder()
+    builder.serviceHttpUrl(config.adminUri).build()
+  }
+
+  private val outTopic = Topic(s"persistent://${config.namespace.asString}/James-${name.asString()}")
+  private val scheduledTopic = Topic(s"persistent://${config.namespace.asString}/${name.asString()}-scheduled")
```

Comment:
Likely a stupid question but why using a separate topic for delayed messages?

My take is that it limits the out-of-order impact and yields better performance.

Do you confirm?